samples/connectedhomeip/light_bulb/src/LightBulbPublishService.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include "LightBulbPublishService.h"
+
+#include "AppTask.h"
+
+#include <logging/log.h>
+#include <zephyr.h>
+
+#include <openthread/message.h>
+#include <openthread/udp.h>
+#include <platform/OpenThread/OpenThreadUtils.h>
+
+#include <inet/InetConfig.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <transport/TransportMgr.h>
+#include <transport/raw/UDP.h>
+
+LOG_MODULE_DECLARE(app);
+
+// Pointer to the LightBulbPublishService instance needed by static
+// PublishTimerHandler method
+static LightBulbPublishService* sLightBulbPublishService;
+
+const char PUBLISH_SERVICE_PAYLOAD[] = "chip-light-bulb";
+const char PUBLISH_DEST_ADDR[] = "ff02::1";
+
+using namespace ::chip;
+using namespace ::chip::Inet;
+using namespace ::chip::Transport;
+using namespace ::chip::DeviceLayer;
+
+void LightBulbPublishService::Init()
+{
+    // Set pointer to the current instance
+    sLightBulbPublishService = this;
+    mIsServiceStarted = false;
+
+    // Initialize timer
+    k_timer_init(&mPublishTimer, &PublishTimerHandler, nullptr);
+    k_timer_user_data_set(&mPublishTimer, this);
+
+    // Initialize UDP endpoint
+    UdpListenParameters udpParams(&chip::DeviceLayer::InetLayer);
+    udpParams.SetAddressType(kIPAddressType_IPv6);
+    udpParams.SetInterfaceId(INET_NULL_INTERFACEID);
+    udpParams.SetListenPort(SERVICE_UDP_PORT);
+    if (mUdp.Init(udpParams))
+    {
+        LOG_ERR("LightBulbPublishService initialization failed");
+    }
+
+    mMessageId = 0;
+    IPAddress::FromString(PUBLISH_DEST_ADDR, mPublishDestAddr);
+}
+
+void LightBulbPublishService::Start(uint32_t aTimeoutInMs, uint32_t aIntervalInMs)
+{
+    if (!mIsServiceStarted)
+    {
+        if (aTimeoutInMs >= aIntervalInMs)
+        {
+            LOG_INF("Started Publish service");
+
+            mPublishInterval = aIntervalInMs;
+            // Calculate number of publish iterations to do considering given timeout and interval values.
+            mPublishIterations = aTimeoutInMs / aIntervalInMs;
+            mIsServiceStarted = true;
+
+            k_timer_start(&mPublishTimer, K_MSEC(aIntervalInMs), K_NO_WAIT);
+        }
+        else
+        {
+            LOG_ERR("Publish service cannot be started: service timeout is greater than interval");
+        }
+    }
+    else
+    {
+        LOG_INF("Publish service is already running");
+    }
+}
+
+void LightBulbPublishService::Stop()
+{
+    if (mIsServiceStarted)
```

Comment:
Actually it's OK to stop an already stopped timer, so this condition may not be checked.

Reply:
I know, but I assumed that it's better to check this condition than execute some part of redundant code.